### ISSUE TITLE
Implementation of Bahdanau attention is possibly different from paper?

### ISSUE THREAD

**Issue by rafaelvalle** [r9y9/tacotron_pytorch#10]
Hi @r9y9.
On the forward pass of the attention RNN, I think you're computing the attention state using the previous attention instead of the current attention.
Page 3 on https://arxiv.org/pdf/1409.0473.pdf

Shouldn't the order of operations be this:

``` python
def forward(self, query, attention, cell_state, memory, processed_memory=None, mask=None, memory_lengths=None):
    
    if processed_memory is None:
        processed_memory = memory
    if memory_lengths is not None and mask is None:
        mask = get_mask_from_lengths(memory, memory_lengths)

    # Compute Alignment (batch, max_time)
    # e_{ij} = a(s_{i-1}, h_j)
    alignment = self.attention_mechanism(cell_state, processed_memory)

    if mask is not None:
        mask = mask.view(query.size(0), -1)
        alignment.data.masked_fill_(mask, self.score_mask_value)

    # Normalize attention weight
    # \alpha_{ij} = softmax(e_{ij})
    alignment = F.softmax(alignment)

    # Attention context vector
    # c_i = \sum_{j=1}^{T_x} \alpha_{ij} h_j
    attention = torch.bmm(alignment.unsqueeze(1), memory)
    attention = attention.squeeze(1)

    # Concat y_{i-1} and c_{i}
    cell_input = torch.cat((query, attention), -1)
    cell_input = cell_input.unsqueeze(1)

    # Feed it to RNN
    # s_i = f(y_{i-1}, c_{i}, s_{i-1})
    cell_output = self.rnn_cell(cell_input, cell_state)

    return cell_output, attention, alignment
```

instead of what we have on this repo right now:

``` python
def forward(self, query, attention, cell_state, memory, processed_memory=None, mask=None, memory_lengths=None):

    if processed_memory is None:
        processed_memory = memory
    if memory_lengths is not None and mask is None:
        mask = get_mask_from_lengths(memory, memory_lengths)

    # Concat y_{i-1} and c_{i}
    cell_input = torch.cat((query, attention), -1)

    # Feed it to RNN
    # s_i = f(y_{i-1}, c_{i-1}, s_{i-1}) should be f(y_{i-1}, c_{i}, s_{i-1}) according to the paper.
    cell_output = self.rnn_cell(cell_input, cell_state)

    # Compute Alignment (batch, max_time)
    # e_{ij} = a(s_{i-1}, h_j)
    alignment = self.attention_mechanism(cell_output, processed_memory)

    if mask is not None:
        mask = mask.view(query.size(0), -1)
        alignment.data.masked_fill_(mask, self.score_mask_value)

    # Normalize attention weight
    # \alpha_{ij} = softmax(e_{ij})
    alignment = F.softmax(alignment)

    # Attention context vector
    # c_i = \sum_{j=1}^{T_x} \alpha_{ij} h_j
    attention = torch.bmm(alignment.unsqueeze(1), memory)

    # (batch, dim)
    attention = attention.squeeze(1)

    return cell_output, attention, alignment
```

**Comment by erogol** [comment]
I this you are right. I was trying to make the attention module work lately. Maybe it is the reason, as I found out the same thing as a possible mistake.

**Comment by erogol** [comment]
I think you dont need this? ```cell_input = cell_input.unsqueeze(1)```

**Comment by rafaelvalle** [comment]
Despite of possibly being different from what is described in the paper, I can confirm that the code in this repo works fine. I want to confirm that the code implemented in this repo is different from the paper and then understand why.

**Comment by r9y9** [comment]
I tried to replicate tensorflow implementation, not the algorithm described in the paper. That would be the possible reason my implementation differs from the original one. Please give me some time to confirm what actually differs.

**Comment by erogol** [comment]
@rafaelvalle I dont know why then, when I run the code, my alignment matrix is not diagonal and full of infinitesimal numbers. Do you have anything you need to enable or disable? Like masking?

**Comment by rafaelvalle** [comment]
@erogol Ryuchi's code runs without any trouble. You might need ~15k iterations to see diagonal like attentions.

**Comment by erogol** [comment]
I checked the TF.contrib but it seems like current implementation is correct. But as you formulate above, the real paper shows your implementation. I am confused :).

On the other side, once I checked the Tacotron paper, the figure shows something similar to TF implementation.

**Comment by rafaelvalle** [comment]
@erogol I think the difference will be marginal but want to make sure the code matches the paper
